analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs

```diff
@@ -36,115 +36,70 @@ public sealed class CollectionEmptinessChecking : SonarDiagnosticAnalyzer
-
         private static readonly DiagnosticDescriptor rule =
             DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
+        private static readonly CSharpExpressionNumericConverter ExpressionNumericConverter = new CSharpExpressionNumericConverter();
+
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax) c.Node;
-                    CheckCountZero(binary.Right, binary.Left, c);
-                    CheckCountOne(binary.Left, binary.Right, c);
-                },
-                SyntaxKind.GreaterThanExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountZero(binary.Left, binary.Right, c);
-                    CheckCountOne(binary.Right, binary.Left, c);
-                },
-                SyntaxKind.LessThanExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountOne(binary.Right, binary.Left, c);
-                    CheckCountZero(binary.Left, binary.Right, c);
-                },
-                SyntaxKind.GreaterThanOrEqualExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountOne(binary.Left, binary.Right, c);
-                    CheckCountZero(binary.Right, binary.Left, c);
-                },
-                SyntaxKind.LessThanOrEqualExpression);
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountZero(binary.Left, binary.Right, c);
-                    CheckCountZero(binary.Right, binary.Left, c);
-                },
-                SyntaxKind.EqualsExpression);
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
                     var binary = (BinaryExpressionSyntax)c.Node;
-                    CheckCountZero(binary.Left, binary.Right, c);
-                    CheckCountZero(binary.Right, binary.Left, c);
+                    var left = ExpressionNumericConverter.TryGetConstantIntValue(binary.Left, out var l_out) ? (int?)l_out : null;
```

Comment:
If `left != null` we don't need to evaluate `TryGetConstantIntValue` for `right`. Let's move the logic to a parametrized method
```
        private static bool CheckExpression(SyntaxNodeAnalysisContext c, ExpressionSyntax expression, bool mirror)
        {
            if (ExpressionNumericConverter.TryGetConstantIntValue(expression, out var value))
            {
                var comparison = CSharpFacade.Instance.Syntax.ComparisonKind(expression.Parent);
                if (mirror)
                {
                    comparison = comparison.Mirror();
                }
                if (comparison.Count(value).EmptyOrNotEmpty() && TryGetCountCall(expression, c.SemanticModel, out var location, out var typeArgument))
                {
                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, location, typeArgument));
                }
                return true;
            }
            else
            {
                return false;
            }
        }
```
and invoke it like this
```
                    if (!CheckExpression(c, binary.Left, true))
                    {
                        CheckExpression(c, binary.Right, false);
                    }
```
I didn't try it, there might be some minor fixes needed.

Reply:
We'll probably need something like `countExpression` / `valueExpression` arguments to make this work because my simple refactoring just uses one of the binary nodes needed.

Reply:
I did a bit different, but with the result that I do not check right, once left has a value.

Reply:
I'd still prefer the parametrized version. We have 3 times the same condition and dual processing.

Reply:
@pavel-mikula-sonarsource I'd further improved it. I think (and hope) you will like this one.

Reply:
Yep, nice